### DISCUSSION TITLE
Binaries included but bunch of compile errors

### DISCUSSION THREAD

**Discussion by SebBil** [alandefreitas/matplotplusplus#62] [Q&A] [answered]
Hi all,

i try to run the matplotplusplus in a windows environment and under visual studio 2017. All includes are correct, but i get a bunch of compile errors. Are there some other dependencies that i have to install first? 

![image](https://user-images.githubusercontent.com/49487453/93626571-e69a8480-f9e3-11ea-96d8-9e4e6bc49c8a.png)

This is only a part of all 1700 errors.
I also tried to change the c++ language to C++17, but also with no luck and around 60 errors

![image](https://user-images.githubusercontent.com/49487453/93627174-cfa86200-f9e4-11ea-9473-c1b4daca6db3.png)

Can i try something else?

**Comment by alandefreitas** [comment]
For what reason is it indicating that a ";" is missing in common.h? Why would just "const" be a syntax error in C++17? All that seems very weird.

Maybe you can follow the instructions for building from source on your computer to figure where things are going differently.
It seems like there's something wrong in the compiler settings. 

The library is building and installing fine for windows, linux, and mac os:
https://github.com/alandefreitas/matplotplusplus/actions?query=workflow%3AMatplotplusplus
Maybe, you can also use that as a reference for the commands to compile from the source code in a clean machine.

**Comment by SebBil** [comment]
Thanks for your response,

i have tried to build the from source, (i'm not very familiar with cmake :( ) but now i get an error on the world_map_10m.cpp, that the expression was not evaluated to a constant.

What i have done:
Cloned the repo
Configure with: cmake .. -G "Visual Studio 15 2017" -A x64 -DCMAKE_CXX_FLAGS="/O2" -DCMAKE_BUILD_TYPE=Release
And tried to build:
PS D:\Projects\matplotplusplus\build> cmake --build . -j 2 --config Release
Microsoft (R)-Build-Engine, Version 15.9.21+g9802d43bc3 für .NET Framework
Copyright (C) Microsoft Corporation. Alle Rechte vorbehalten.

  Checking Build System
  Building Custom Rule D:/Projects/matplotplusplus/source/3rd_party/CMakeLists.txt
  algebra.cpp
  fruchterman_reingold.cpp
  kamada_kawai.cpp
  layout.cpp
  nodesoup.cpp
  Code wird generiert...
  nodesoup.vcxproj -> D:\Projects\matplotplusplus\build\source\3rd_party\Release\nodesoup.lib
  Building Custom Rule D:/Projects/matplotplusplus/source/matplot/CMakeLists.txt
  backend_interface.cpp
  gnuplot.cpp
  backend_registry.cpp
  axes_type.cpp
  Building Custom Rule D:/Projects/matplotplusplus/test/generate_examples/CMakeLists.txt
  main.cpp
D:\Programme\Microsoft Visual Studio\2017\Enterprise\VC\Tools\MSVC\14.16.27023\include\xlocale(319): warning C4530: C++-Handler verwendet, aber Entladesemantik is
t nicht aktiviert. Geben Sie /EHsc an. [D:\Projects\matplotplusplus\build\test\generate_examples\generate_examples.vcxproj]
  axes_object.cpp
  generate_examples.vcxproj -> D:\Projects\matplotplusplus\build\test\generate_examples\Release\generate_examples.exe
  axis_type.cpp
  figure_type.cpp
  figure_registry.cpp
  legend.cpp
  line_spec.cpp
  colors.cpp
  common.cpp
  contourc.cpp
  world_cities.cpp
  world_map_10m.cpp
D:\Projects\matplotplusplus\source\matplot\util\world_map_10m.cpp(16): error C2131: Ausdruck wurde nicht zu einer Konstanten ausgewertet. [D:\Projects\matplotplus
plus\build\source\matplot\matplot.vcxproj]
D:\Projects\matplotplusplus\source\matplot\util\world_map_10m.cpp(414749): error C2131: Ausdruck wurde nicht zu einer Konstanten ausgewertet. [D:\Projects\matplot
plusplus\build\source\matplot\matplot.vcxproj]
  world_map_50m.cpp
  world_map_110m.cpp
  bars.cpp
  box_chart.cpp
  circles.cpp
  Code wird generiert...
  Kompilieren...
  contours.cpp
  error_bar.cpp
  filled_area.cpp
  function_line.cpp
  histogram.cpp
  labels.cpp
  line.cpp
  matrix.cpp
  network.cpp
  parallel_lines.cpp
  stair.cpp
  string_function.cpp
  surface.cpp
  vectors.cpp
  axes_functions.cpp
  histcounts.cpp
  Code wird generiert...

**Comment by alandefreitas** [comment, marked as answer]
Can you try that with:

```bash
mkdir build
cmake -version
cmake .. -G "Visual Studio 16 2019" -A x64 -DCMAKE_CXX_FLAGS="/O2" -DCMAKE_BUILD_TYPE=Release
cmake --build . -j 2 --config Release
cmake --install .
cpack .
```

This is how we build it in [Continuous Integration for Windows](https://github.com/alandefreitas/matplotplusplus/runs/1140717168?check_suite_focus=true). 

To build from source you need to "Make sure you have a recent version of [Visual Studio](https://visualstudio.microsoft.com/)" and this is very true. I don't think "Visual Studio 15 2017" implemented the complete C++17 standard yet.

**Reply by SebBil** [reply]
That was the problem.

Another hint: I also included some other libs, and the compile errors in the common header are because of another include that define min and max, so i have to undefine this two macros and the code works,

Thanks a lot for your help.

**Reply by alandefreitas** [reply]
What other libs did you need to include?

I'm not sure what to do about those macros sometimes. There are lots of hacks in the build scripts to go around these problems with MSVC. The min macro is [one of them](https://github.com/alandefreitas/matplotplusplus/blob/7864cc34984362c70ce512bb8c0c5fa6586b3ca1/source/matplot/CMakeLists.txt#L118). If you import matplot++ with cmake it should define NOMINMAX for you so that you don't have this problem. 

But if you're importing the library without CMake, there's no easy workaround to make sure the user has defined NOMINMAX outside matplot.h. It seems like this problem is so recurrent that it's something of which people using MSVC will need to be aware for any external library.
